src/rust/engine/process_execution/src/docker.rs

```diff
@@ -33,6 +38,78 @@ pub struct CommandRunner {
   named_caches: NamedCaches,
   immutable_inputs: ImmutableInputs,
   keep_sandboxes: KeepSandboxes,
+  image_pull_cache: Mutex<BTreeMap<String, Arc<OnceCell<()>>>>,
+  image_pull_policy: ImagePullPolicy,
+}
+
+#[allow(dead_code)] // TODO: temporary until docker command runner is hooked up
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum ImagePullPolicy {
+  Always,
+  IfMissing,
+  Never,
+  OnlyIfLatestOrMissing,
```

Comment:
It's probably worth noting somewhere that given the cache, the policy will only be applied once per-image + per-`pantsd` instance.

That's probably fine, but probably makes it less useful to provide a lot of configurability here.

Reply:
Yeah, which would make the "always" policy not actually "always." The `image_pull_cache` is meant to debounce concurrent pull attempts. I'll need to iterate on this.

Reply:
The `Context` passed into the `CommandRunner` has a bit of session info, so could maybe make it more intelligent using that facility.

But this is a lower priority than getting caching in place, so keeping things simple until we need it seems like the right approach.

Reply:
Is `Context` created once per "session"? If so, I could just move`image_pull_cache` there and that could take care of the issue.

Reply:
The `Context` is created once per process: https://github.com/pantsbuild/pants/blob/1f24bcdbeeddc1847bdc47b31fcbba314323cb2a/src/rust/engine/src/nodes.rs#L449-L453

...but it contains ids which are updated either once for the entire run (`build_id`), or once per iteration of a `--loop` (`run_id`): https://github.com/pantsbuild/pants/blob/1f24bcdbeeddc1847bdc47b31fcbba314323cb2a/src/rust/engine/src/session.rs#L72-L108

Reply:
Then the code could probably use `build_id` changing to know when to clear the image pull cache which would have the effect of allowing "always" pulls to occur again while still debouncing the pull.

Reply:
Cache is now cleared when `build_id` changes.